# Halobacterium salinarum proteome dynamics in the context of mRNA expression

This repository stores the scripts used to perform the following tasks:  

* Exploratory analyses for proteomics datasets analyzed by OneOmics and Spectronaut.
* Plots.

Brief description of each file stored in scripts folder:

* **0_loadingLibs.R**: Loads all libraries required to run the entire analysis.
* **1_locusTagDict.R**: Create a dictionary of locus_tags. This is required to integrate the proteomics dataset to the gene expression dataset. To improve the number of matched entries, we use an approach based in Reciprocal Best BLAST-Hit.
* **2_parseOneOmicsResults.R**: Parse and filter proteomics results; Plot and save the volcano plot and the heat map for the OneOmics dataset.
* **3_PlotPCAOneOmics.R**: Parse PCA data generated by OneOmics and plot/save the figure.
* **4_parseSpectronautResults.R**: Read and parse abundance data provided by Spectronaut; Generate and save the heat map for this dataset.
* **5_PlotPCASpectronaut.R**: Wrangle Spectronaut dataset and run PCA followed by its plot.
* **6_RNADEanalysis.R**: Read kallisto count data generated by [runKallisto pipeline](https://github.com/alanlorenzetti/runKallisto) and perform differential expression analysis.
* **7_parseAndGenScatters.R**: Parse protein and mRNA levels and relative changes and generate scatter plots between those two variables.

This set of scripts is powered by `RStudio`, `tidyverse`, and `Bioconductor`.
